File: 15-3sum/15-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        
        nums.sort()
        ans = []
        
        print(nums)
        
        size = len(nums)
        for p1 in range(size-2):
            
            
            if (p1 >0 and nums[p1] == nums[p1-1]):
                continue
                
            n1 = nums[p1]
            
            p2 , p3 = p1+1 , size-1
            
            while(p2 < p3):
                
                curr_sum = nums[p1] + nums[p2] + nums[p3]

                if (curr_sum == 0):
                    
                    ans.append([nums[p1] , nums[p2] , nums[p3]])
                    
                    p2 +=1
                    while(p2 < p3 and nums[p2] == nums[p2-1]):
                        p2 +=1
                    
                    p3 -=1
                    while(p3 >=0 and nums[p3] == nums[p3+1]):
                        p3 -=1
                
                elif (curr_sum <0):
                    p2 +=1
                
                else:
                    p3 -=1
        
        return (ans)
                

File: 15-3sum/15-3sum.rs
impl Solution {
    pub fn three_sum(mut nums: Vec<i32>) -> Vec<Vec<i32>> {
    
    let mut ans = vec![];

    nums.sort();

    let size = nums.len();
    if size <3{
        return ans;
        }

    let mut low = 0;
    let mut high = 0;

    for idx in 0..size-2{

        if (idx>0 && nums[idx] == nums[idx-1]){
            continue;
        }
        let curr_ele = nums[idx];

        low = idx+1;
        high = size-1;

        while low < high{

            let sum = curr_ele + nums[low] + nums[high];

            if sum == 0 {
                let mut res = vec![nums[idx], nums[low] , nums[high]];
                ans.push(res);

                low +=1;
                high -=1;

                while low < high && nums[low] == nums[low-1]{
                    low = low+1;
                }

                while high > low && nums[high] == nums[high+1]{
                    high -=1;
                }
            }

            else if sum > 0{
                high -=1;
            }
            else{
                low +=1
            }
        }

    }

    return ans;
}
}